src/core/MIGRATION.md

```diff
@@ -669,27 +670,94 @@ No. That said, the migration process will require a lot of refactoring, and Type
+
+#### How to decide what code can be statically imported
+
+The general rule of thumb here is: any module that is not purely functional should not be shared statically, and instead should be exposed at runtime via the plugin's `setup` and/or `start` contracts.
+
+When deciding whether or not to share via exports or plugin contracts, consider these questions:
+- Is it's behavior dependent on any state inside the module?
+- If a plugin uses an old copy of this module, will it still break?
+
+If you answered yes to any of the above questions, you probably have an impure module that cannot be shared across plugins.
+
+Examples of code that could be shared statically:
+- Constants. Strings and numbers that do not ever change (even between Kibana versions)
+- React components. Make sure these components are not dependent on or pre-wired to Core services. In many of these cases you can export a HOC that takes the Core service and returns a component wired up to that particular service instance.
+- Pure computation functions.
+
+Examples of code that could **not** be shared statically:
+- A function that queries Elasticsearch, but does not take an ES Client as a parameter.
+- A React component that reads `uiSettings` from a variable defined in the module.
+- A function that allows other plugins to register values that get pushed into an array defined internally to the module.
+
+#### Concrete Example
+
+Ok, you've decided you want to export static code from your plugin, how do you do it? In principle, the New Platform only considers values exported from `my_plugin/public` and `my_plugin/server` to be stable. Our tooling will enforce that these APIs do not break between major versions (in the future). All code shared among plugins should be exported in these modules like so:
+
+```ts
+// my_plugin/public/index.ts
+export { MyPureComponent } from './components';
+
+// regular plugin export used by core to initialize your plugin
+export const plugin = ...;
+```
+
+These can then be imported using relative paths from other plugins:
+
+```ts
+// my_other_plugin/public/components/my_app.ts
+import { MyPureComponent } from '../my_plugin/public';
+```
+
+If you have code that should be available to other plugins on both the client and server, you can have a common directory. _See [How is "common" code shared on both the client and server?](#how-is-common-code-shared-on-both-the-client-and-server)_
+
+### How can I avoid passing Core services deeply within my UI component tree?
+
+There are some Core services that are purely presentational, for example `core.overlays.openModal()` or `core.application.createLink()` where UI code does need access to these deeply within your application. However, passing these services down as props throughout your application leads to lots of boilerplate. To avoid this, you have two options:
+- Use React Context to provide these services to large parts of your React tree; or
```

Comment:
optional:
as we reference below on Redux I'd add here info about injecting those runtime deps in data layer context:
- redux-thunk `thunk.withExtraArgument({ http }))` https://github.com/reduxjs/redux-thunk#injecting-a-custom-argument
- redux-saga `createSagaMiddleware ({ context: http })` https://redux-saga.js.org/docs/api/#createsagamiddlewareoptions